### PR TITLE
Add advanced data structure and some mathematical algorithms; included Java in A* implementation.

### LINKED ISSUE #602: Implementation of Advanced Data Structures, Number Theory, and Advanced Search Algorithms
### 1. Advanced Data Structures
- **Disjoint Set Union (Union-Find)**: Efficient data structure to track the union of disjoint sets.
- **Segment Trees**: Used for efficient range queries.
- **Fenwick Tree (Binary Indexed Tree)**: A more space-efficient data structure for cumulative frequency tables.

### 2. Number Theory Algorithms
- **Greatest Common Divisor (GCD) and Least Common Multiple (LCM)**: Fundamental algorithms for divisibility.
- **Sieve of Eratosthenes**: Efficient algorithm to find all prime numbers up to a given limit.
- **Modular Arithmetic**: Core operations and algorithms for modularity, which are essential in cryptography and number theory.

### 3. Combinatorial Algorithms 
- Combinatorial algorithms such as **Permutations**, **Combinations**, and **Binomial Coefficients** with implementations.

### 4. Advanced Search Algorithms
A new section dedicated to advanced search techniques:
- **A * Search Algorithm**: A powerful pathfinding and graph traversal algorithm widely used in AI and game development.

These updates will ensure a more comprehensive collection of data structures and algorithms for all learners and developers.

### PR DESCRIPTION
**Pull request by ayan0xdl**
### Description
This pull request adds advanced data structures and algorithms, implements several mathematical algorithms, and includes Java code along with time and space complexity for the A* search algorithm. These enhancements aim to improve the project's coverage of important algorithms and provide clearer implementations for better understanding.

Fixes #602 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by ayan0xdl** [comment]
could you pls check @ajay-dhangar and if possible give me level2 label

File: docs/mathematics-algorithms/sieve-of-eratosthenes.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+## Algorithm
+
+### Steps:
+1. Create a list of consecutive integers from 2 to \( n \) (the limit).
+2. Start with the first prime number \( p = 2 \).
+3. Mark all multiples of \( p \) (from \( p^2 \) to \( n \)) as composite (not prime).
+4. Find the next number in the list that is not marked. This number is the next prime \( p \).
+5. Repeat steps 3 and 4 until \( p^2 \) is greater than \( n \).
+6. The numbers that remain unmarked in the list are all the prime numbers up to \( n \).
+
```
Comment (ajay-dhangar) [inline docs/mathematics-algorithms/sieve-of-eratosthenes.md]: replace `\( n \)` with `$n$`, `\( p = 2 \)` with `$p=2$`, `\( p^2 \)` with `$ p^2$`, and `\( p \)` with `$p$`

File: docs/mathematics-algorithms/sieve-of-eratosthenes.md [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+- Start with the list: [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30]
+- Mark multiples of 2: [2, 3, X, 5, X, 7, X, 9, X, 11, X, 13, X, 15, X, 17, X, 19, X, 21, X, 23, X, 25, X, 27, X, 29, X]
+- Mark multiples of 3: [2, 3, X, 5, X, 7, X, X, X, 11, X, 13, X, X, 17, X, 19, X, X, X, 23, X, 25, X, X, X, 29, X]
+- Continue this process until all multiples of primes are marked.
+- The unmarked numbers are: 2, 3, 5, 7, 11, 13, 17, 19, 23, 29.
+
+## Time Complexity
+
+- The time complexity of the Sieve of Eratosthenes is \( O(n log(log(n))) \), where \( n \) is the limit up to which prime numbers are generated. This makes it very efficient for large values of \( n \).
+
```
Comment (ajay-dhangar) [inline docs/mathematics-algorithms/sieve-of-eratosthenes.md]: You need to check out https://docusaurus.io/docs/markdown-features/math-equations for more details on using math equations in our documentation.

File: docs/mathematics-algorithms/sieve-of-eratosthenes.md [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+## Space Complexity
+
+- The space complexity is \( O(n) \) due to the array used to mark the prime numbers.
```
Comment (ajay-dhangar) [inline docs/mathematics-algorithms/sieve-of-eratosthenes.md]: also, replace `\( O(n) \)` with `$O(n)$`

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
Your content is very good. However, based on our suggestions, you need to correct a few mistakes.

**Comment by ayan0xdl** [comment]
made the changes according to your suggestions pls review it and also assign me level 2 label in the issue as well @ajay-dhangar

**Review by ajay-dhangar — APPROVED** [review]
Done

**Pull request merged**